src/Field/BaseFieldDescription.php

```diff
@@ -0,0 +1,326 @@
+
+    /**
+     * @var array[] cached object field getters
+     */
+    private static $fieldGetters = [];
+
+    public function setFieldName(string $fieldName): void
+    {
+        $this->fieldName = $fieldName;
+    }
+
+    public function getFieldName(): ?string
+    {
+        return $this->fieldName;
+    }
+
+    public function setName(string $name): void
+    {
+        $this->name = $name;
+
+        if (!$this->getFieldName()) {
+            $this->setFieldName(substr(strrchr('.'.$name, '.'), 1));
+        }
+    }
+
+    public function getName(): ?string
+    {
+        return $this->name;
+    }
+
+    public function setTemplate(string $template): void
+    {
+        $this->template = $template;
+    }
+
+    public function getTemplate(): ?string
+    {
+        return $this->template;
+    }
+
+    /**
+     * @param int|string $type
+     */
+    public function setType($type): void
+    {
+        $this->type = $type;
+    }
+
+    public function getType()
+    {
+        return $this->type;
+    }
+
+    /**
+     * @param mixed $default
+     *
+     * @return mixed
+     */
+    public function getOption(string $name, $default = null)
+    {
+        return isset($this->options[$name]) ? $this->options[$name] : $default;
+    }
+
+    /**
+     * @param mixed $value
+     */
+    public function setOption(string $name, $value): void
+    {
+        $this->options[$name] = $value;
+    }
+
+    public function setOptions(array $options): void
+    {
+        // set the type if provided
+        if (isset($options['type'])) {
+            $this->setType($options['type']);
+            unset($options['type']);
+        }
+
+        // set the template if provided
+        if (isset($options['template'])) {
+            $this->setTemplate($options['template']);
+            unset($options['template']);
+        }
+
+        $this->options = $options;
+    }
+
+    public function getOptions(): array
+    {
+        return $this->options;
+    }
+
+    public function mergeOption(string $name, array $options = []): void
+    {
+        if (!isset($this->options[$name])) {
+            $this->options[$name] = [];
+        }
+
+        if (!\is_array($this->options[$name])) {
+            throw new \RuntimeException(sprintf('The key `%s` does not point to an array value', $name));
+        }
+
+        $this->options[$name] = array_merge($this->options[$name], $options);
+    }
+
+    public function mergeOptions(array $options = []): void
+    {
+        $this->setOptions(array_merge_recursive($this->options, $options));
+    }
+
+    public function getAssociationMapping(): array
+    {
+        return $this->associationMapping;
+    }
+
+    public function getFieldMapping(): array
+    {
+        return $this->fieldMapping;
+    }
+
+    public function getParentAssociationMappings(): array
+    {
+        return $this->parentAssociationMappings;
+    }
+
+    /**
+     * @param int|string $mappingType
```

Comment:
Do we need this doc block? Ain't it the same on the interface?

Reply:
Isn't easier to have the phpdoc in the file you're reading when you're coding/looking in your vendor, etc ?
I think it's always useful

Reply:
Most IDEs can use autocompletion / show methods docs from upper interfaces. The problem is that duplicate docblocks will one day get out of sync.

Reply:
I understand. 

But because there was no docblocks in the same class in the SonataAdminBundle, a bug was easily introduce in master by restricting the value to `?string`.
https://github.com/sonata-project/SonataAdminBundle/blob/master/src/Admin/BaseFieldDescription.php#L227

You proved me that I made the right things by adding this tricky docblock: The first time you read it, you thought that `int` was a mistake.
Without this docblock I can bet that someone gonna restrict the value to `string`.